### PR TITLE
Run Windows arm32 tests and publish assets

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/corefx/issues/38863

Prepare yourself for some errors in the leg 😁 - as it was never tested.

cc @safern @MattGal

### CODE REVIEW

**Comment by safern** [comment]
Why adding it for PRs? Do we want to do that? How slow is that leg, because for linux we where hitting a lot of timeouts.

**Comment by ViktorHofer** [comment]
> Why adding it for PRs? Do we want to do that?

If coreclr is going to add it to their PRs we want to do the same. (We currently don't do it for arm64 because of resource constraints)

> How slow is that leg, because for linux we where hitting a lot of timeouts.

That's what this PR will tell :)

**Comment by BruceForstall** [comment]
coreclr will only add it as outerloop, not for PR. So I think it's ok if it's only for outerloop (or official build) in corefx as well.

**Review by BruceForstall — APPROVED** [review]
The changes look good to me. However, as I wrote, I don't think you need the PR testing unless corefx wants it. Coreclr only needs the "_publishTests: true" Release version, which it appears is only outerloop.

**Comment by safern** [comment]
I think we should start with rolling builds. So its fine to test on this PR but please do change it only for fullMatrix and then we can move it to PRs.

> That's what this PR will tell :)

I don't think that is realistic because that will be the only payload. The problem is when the payload starts to be bigger (i.e: multiple PRs open + rolling builds + outerloop). cc: @MattGal thoughts?

**Comment by BruceForstall** [comment]
I spoke to @RussKeldorph offline and he urges you to _not_ add Windows/arm32 PR testing at this time, based on current machine availability.

**Comment by RussKeldorph** [comment]
I recommend running ARM and ARM64 testing only in official builds or less frequently.  We are *extremely* resource-constrained on these platforms, especially Windows.  I suspect that, unlike CoreCLR, very few CoreFX changes introduce ARM-specific issues.  The one exception would be CoreCLR ingestions, so it would be nice to be able to trigger ARM testing for those PRs and a handful of other risky ones, but the cost/benefit of running these on every PR or commit just isn't there IMO.

**Comment by ViktorHofer** [comment]
Thanks for the feedback, that makes perfect sense. I didn't know that coreclr isn't running them in "innerloop" either. I will remove that from the yaml after I diagnosed the test failures as currently all work items are failing.

@MattGal can you tkae a look at the Windows arm_Debug errors? They are very cryptic:

> Executed on ddvsohumw013
The system cannot find the path specified.

**Comment by MattGal** [comment]
@ViktorHofer uh oh I think I know what this is (machine setup issue).  Let me do a quick investigation and we'll get the DDFUN folks looking at it.

**Comment by MattGal** [comment]
@ViktorHofer the problem is what I thought, it's got a very old copy of the Helix python scripts from before the "CorrelationPayloadUrisWithDestinations" feature was added, so it's not actually unpacking correlation payloads.  I'm chatting with @parose1 now about how soon they might get updated.

**Comment by MattGal** [comment]
@ViktorHofer - fixing this is tracked by https://github.com/dotnet/core-eng/issues/6812.  Sorry for the inconvenience, we've been reinventing the way those machines get setup and they were set up in a time before that happened.

**Comment by parose1** [comment]
I have completed https://github.com/dotnet/core-eng/issues/6812 and all ARM32 systems are now updated to python3 with the new scripts. New systems added will have this fix and should not run into this issue. @ViktorHofer please let me know if you are still seeing any issues. Thanks

**Comment by ViktorHofer** [comment]
Great! Do I need to update the docker versions in the yml?

**Comment by ViktorHofer** [comment]
This isn't blocked anymore. The test leg is now running correctly. Seems we have a non-trivial amount of tests to fix, I haven't yet investigated if those failures are product issues.

cc @danmosemsft @stephentoub

**Comment by ViktorHofer** [comment]
99% of the failures are related to gdiplus.dll not available on Windows arm. cc @safern

**Comment by ViktorHofer** [comment]
Seems that at least some failures are product related (ie Console) and should be fixed for 3.0 as arm32 is a supported configuration.

@danmosemsft we should prioritize this item correctly

**Comment by danmosemsft** [comment]
@joperezr did a manual run of CoreFX tests on Windows ARM32 6 months ago ([these](https://github.com/dotnet/corefx/issues?q=is%3Aopen+is%3Aissue+label%3Aarch-arm32)). @joperezr can you please see whether there are new failures you didn't log before, and check we have the right stuff if anything in 3.0?

**Comment by danmosemsft** [comment]
BTW I don't see any Console failures (?) in the rsults
https://dev.azure.com/dnceng/public/_build/results?buildId=238718&view=ms.vss-test-web.build-test-results-tab&runId=6076400&resultId=100322&paneView=debug

Of those, the ones that likely need investigation are the EventSource ones and the Environment ones. Others seem likely OS quirks etc that aren't interesting. @joperezr can look hopefully.

**Comment by BruceForstall** [comment]
MC shows a Console failure: https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F38906~2Fmerge/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190625.59/workItem/System.Console.Tests/analysis/xunit/ReadAndWrite~2FOutputEncodingPreamble

**Comment by danmosemsft** [comment]
That's interesting. @ViktorHofer any idea why AzDo: https://dev.azure.com/dnceng/public/_build/results?buildId=238718&view=ms.vss-test-web.build-test-results-tab does not show all the tests that MC does here https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F38906~2Fmerge/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190625.59 ?

**Comment by ViktorHofer** [comment]
cc @alexperovich who should know

**Comment by MattGal** [comment]
@ViktorHofer / @danmosemsft I'll let Alex throw in some more details if he likes but at a glance, the problem is that the python "Vsts" library isn't getting properly installed / put on the PYTHONPATH on some of the ARM32 machines (@parose1  FYI).  There doesn't seem to be evidence of the test reporter script trying to install it, I'll chat with @alexperovich when I can and see if this is expected to already be on the machines:

```
Traceback (most recent call last):
  File "C:\Data\helix\work\47387781-e2a5-4cdb-a59e-730f4ed5e222\Payload\reporter\run.py", line 9, in <module>
    from azure_devops_result_publisher import AzureDevOpsTestResultPublisher
  File "C:\Data\helix\work\47387781-e2a5-4cdb-a59e-730f4ed5e222\Payload\reporter\azure_devops_result_publisher.py", line 4, in <module>
    from vsts.vss_connection import VssConnection
ModuleNotFoundError: No module named 'vsts'
```
The machine it ran on was imaged on 6/26 so it's possible something's off there but I don't see any specific evidence of that, just that this work item tried to import something that wasn't pip-install'ed first.

One possibility is that, like the docker library, this simply isn't available on Windows ARM32 Python  (which is an exotic and weird flavor).  I'll try this experiment shortly.

Other semi-interesting stuff:
- This work item timed out while uploading to AzDO which will cause indeterminate behavior of some sort: https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F38906~2Fmerge/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190625.59/workItem/System.Runtime.Tests/wilogs
- AzDO has been working to mitigate problems with our test results filling up their database, and slowness / tarpitting can still happen with the uploads.

**Comment by MattGal** [comment]
One other data point;  The VSTS results here are legitimately wrong (if not just extremely misleading).  The linked AzDO runs show failures for System.ComponentModel.TypeConverter tests on ARM32:

image

But we see from the MC link that it had the same problem the other ARM32 runs had:

https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F38906~2Fmerge/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190625.59/workItem/System.ComponentModel.TypeConverter.Tests/wilogs

```
=== TEST EXECUTION SUMMARY ===
   System.ComponentModel.TypeConverter.Tests  Total: 6904, Errors: 0, Failed: 0, Skipped: 0, Time: 15.925s
----- end Fri 06/28/2019 14:24:39.62 ----- exit code 0 ----------------------------------------------------------
 Fri 06/28/2019-14:24:39.63
 Fri 06/28/2019-14:24:39.81
Traceback (most recent call last):
  File "C:\Data\helix\work\47387781-e2a5-4cdb-a59e-730f4ed5e222\Payload\reporter\run.py", line 9, in <module>
    from azure_devops_result_publisher import AzureDevOpsTestResultPublisher
  File "C:\Data\helix\work\47387781-e2a5-4cdb-a59e-730f4ed5e222\Payload\reporter\azure_devops_result_publisher.py", line 4, in <module>
    from vsts.vss_connection import VssConnection
ModuleNotFoundError: No module named 'vsts'
```

... plus had no failures.   I filed https://github.com/dotnet/core-eng/issues/6845 to do a more thorough investigation.

**Comment by ViktorHofer** [comment]
Does anybody know why https://www.nuget.org/packages/runtime.native.System.Data.SqlClient.sni/4.6.0-preview6.19303.8 doesn't have a arm32 dependency? I believe that's the package: https://www.nuget.org/packages/System.Data.SqlClient.sni-arm/4.0.0-beta-23225

cc @divega @roji

**Comment by divega** [comment]
> Does anybody know why https://www.nuget.org/packages/runtime.native.System.Data.SqlClient.sni/4.6.0-preview6.19303.8 doesn't have a arm32 dependency? I believe that's the package: https://www.nuget.org/packages/System.Data.SqlClient.sni-arm/4.0.0-beta-23225

We would need to ask @David-Engel or @saurabh500.

**Comment by divega** [comment]
@ViktorHofer adding arm32 support is tracked as a feature in https://github.com/dotnet/SqlClient/issues/117. See https://github.com/dotnet/corefx/issues/33999 for a previous discussion about this.

**Comment by joperezr** [comment]
> @joperezr did a manual run of CoreFX tests on Windows ARM32 6 months ago (these). @joperezr can you please see whether there are new failures you didn't log before, and check we have the right stuff if anything in 3.0?

@danmosemsft yes I will sync up with @ViktorHofer and make sure that there are no new failures, and if there are, we'll log issues for them.

**Comment by ViktorHofer** [comment]
> is this run running innerloop + outerloop? Back 6 months ago when we originally did this we only ran innerloop.

Just innerloop

**Comment by BruceForstall** [comment]
@ViktorHofer What's the status here? Looks like you're going through fixing/disabling tests that don't work on Windows/arm32?

**Comment by ViktorHofer** [comment]
Right. There are about 70 tests left that need to be conditioned...

**Comment by BruceForstall** [comment]
@ViktorHofer Any progress here?

**Comment by ViktorHofer** [comment]
I have one item with a higher prio on my list and will get back to this PR asap.

File: eng/pipelines/windows.yml [hunk 1]
```diff
@@ -50,6 +50,12 @@ jobs:
               _architecture: x64
               _framework: uap
               _helixQueues: $(uapNetfxQueues)
+
+            arm_Debug:
+              _BuildConfig: Debug
+              _architecture: arm
+              _framework: netcoreapp
+              _helixQueues: $(windowsArmQueue)
```
Comment (ericstj) [inline eng/pipelines/windows.yml]: Plan is to remove these lines before merging since we don't have enough hardware to run this on every PR.

**Comment by eiriktsarpalis** [comment]
@ViktorHofer Friendly nudge to follow up on PR status

**Comment by ViktorHofer** [comment]
Fixing the remaining failures this week.

**Comment by ViktorHofer** [comment]
Just wanted to let you know that I'm currently fully booked with other more important stuff so this has to wait probably a few more weeks.

**Comment by stephentoub** [comment]
@ViktorHofer, shall we close it then?  Presumably you'll pick it up in the consolidated repo and it would get closed anyway?

**Comment by ViktorHofer** [comment]
I would prefer to keep this open for one or two more weeks. @safern or I might have some resources later this or next week. Feel free to close if no progress was made until EOM.

**Comment by ViktorHofer** [comment]
I reconsidered and will close this until I have more time for it.

**Pull request closed without merging**